kv-cache : avoid modifying recurrent cells when setting inputs

compilade opened the pull request:
~~**NOTE**: this targets #13746, **not** `master`.~~ (`master` is now the base branch of this PR since #13746 was merged)

@ggerganov As discussed in <https://github.com/ggml-org/llama.cpp/pull/9126#issuecomment-2913159010>, this ports some of the changes from #9126 to #13746 for recurrent caches.

It mostly works, but there is still something wrong somewhere indicated by non-consecutive
token positions when running mamba-130m-hf with llama-parallel:

```console
$ llama-parallel -m ./mamba-130m/ggml-model-f16.gguf -np 5 -ns 8
...
find_slot: non-consecutive token position 0 after 335 for sequence 1 with 1 new tokens
...
find_slot: non-consecutive token position 0 after 346 for sequence 3 with 1 new tokens
...
```

This was not a problem in #9126, so it might (or not) relate to other recent changes in how the kv cache is handled.

I'll attempt to figure out what is wrong by updating #9126 to the latest `master` and see if the problem also appears (EDIT: **it does cause the problem to appear**).

Is there any recent change in how `kv_cell.pos` is handled which comes to mind and could cause this? (`pos` is not reset properly between re-uses of clients in `parallel`, at least for recurrent models)

I'm suspecting #13598 might be related.

EDIT: it seems like with `-pps`, the problem isn't there, and this was the default behavior before #13598.

---

*Make sure to read the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

compilade committed:
kv-cache : avoid modifying recurrent cells when setting inputs

* kv-cache : remove inp_s_mask

It was replaced with equivalent and simpler functionality
with rs_z (the first zeroed state) and the already-existing inp_s_copy.

* kv-cache : fix non-consecutive token pos warning for recurrent models

The problem was apparently caused by how the tail cells were swapped.

* graph : simplify logic for recurrent state copies

* kv-cache : use cell without src refs for rs_z in recurrent cache

ggerganov commented on src/llama-kv-cache.cpp:
```diff
@@ -2310,21 +2308,12 @@ bool llama_kv_cache_recurrent::prepare(const std::vector<llama_ubatch> & ubatche
-    //        break;
-    //    }
-    //}
+    for (const auto & ubatch : ubatches) {
+        if (!find_slot(ubatch)) {
+            success = false;
+            break;
+        }
+    }
```
This is the "preparation phase".

ggerganov replied:
Yes, I confirm that the warnings are still there even without preparation - both on this branch and on the target branch. It's possible that I hallucinated that the warnings disappear without preparation.

> When using -pps the warnings are gone, though. Maybe this was an existing problem which was surfaced by not making -pps the default in parallel.cpp. I'll try to find the source of the kv_cell.pos discrepancy.

You are likely right. Will take a look as well tomorrow.

ggerganov replied:
Btw, one change in #13746 that could be relevant to this is that we now allocate separate ubatch data buffers for each `llama_ubatch`:

https://github.com/ggml-org/llama.cpp/pull/13746/files#diff-e86f4c320ddf096b16dccbc876be6a50f6d6fc4e690b7ebba8a526cd8caa8f14R52-R63

On `master`, the ubatches were simply views of a single buffer, the contents of which were updated on every `split`. But this was not compatible with the peparation logic, so now we allocate separate buffers.

gabe-l-hart commented:
@ggerganov @compilade I'd love to get an update on the state of this branch (and #7727) as we prepare for the Granite 4 launch. It seems like there have been a number of structural changes that need to be accommodated since this branch was last updated to `master`. If it would help, I'd be happy to take a pass at resolving these conflicts and open a PR into this branch.

compilade committed:
Merge branch 'master' into compilade/readonly-recurrent-inputs

gabe-l-hart commented:
Some more evidence that _might_ be helpful as I'm trying to repro the swapped answers mentioned in https://github.com/ggml-org/llama.cpp/pull/13834#discussion_r2110847307. I don't see answers being swapped, but I _do_ see some curious inconsistencies between running with/without `-pps`. My rough understanding suggests that we would expect identical results with and without `-pps` since it's simply attempting to reuse the cache for the system prompt when enabled.

On my machine, the (seeded) random selection of prompts is choosing (including the duplication):

1. What is the meaning of life?
2. If you could have any superpower, what would it be?
3. I want to learn how to play the piano. What would be the best way to do it?
4. I want to learn how to play the piano. What would be the best way to do it?
5. Recommend some interesting books to read.
6. Recommend some interesting books to read.
7. What is the best way to cook a steak?
8. Recommend some interesting books to read.

The commands I'm running are:

```sh
./bin/llama-parallel -m ~/models/mamba-130m-hf/mamba-130M-hf-F16.gguf -np 5 -ns 8 --temp 0 --repeat-penalty 1.1 -ub 2 --junk 0

./bin/llama-parallel -m ~/models/mamba-130m-hf/mamba-130M-hf-F16.gguf -np 5 -ns 8 --temp 0 --repeat-penalty 1.1 -ub 2 --junk 0 -pps
```

What I'm seeing is that for _most_ of the prompts, the responses are identical with / without `-pps`, but the responses for (3) and (7) differ. The curious thing is that they _deterministically_ differ (3: [without] `I like to read.` / [with] `I would suggest you play the piano.`, 7: [without] `The best way to cook a steak is to use the same as in [Section \[sec:model\]]{}.\n<|endoftext|>` / [without] `The best thing about this is that it's not a "new" game. It's an old one, and the new version has some great features like 3D graphics, voice acting, and more.\nIt's also very easy to play, and you can even use your mouse for the controls!<|endoftext|>`). Further when running without `pps`, the prompts for (3) and (4) should be identical, but the responses are not.

I'm not sure if this is actually helpful, but it might help zero in on a specific repro and/or hint at what's getting mixed up.

gabe-l-hart commented:
I just rebuilt with your latest changes and I'm seeing different inconsistencies between with / without `-pps`. In this instance, the order that the sequences get assigned sequence IDs is different between with / without.

I've been walking through running with/without using `lldb` and dumping out the `cell` struct at [llama-kv-cache-recurrent.cpp:L158](https://github.com/ggml-org/llama.cpp/blob/compilade/readonly-recurrent-inputs/src/llama-kv-cache-recurrent.cpp#L158). When I start to see the with / without results diverge, the consistent thing I'm seeing is that the `pos` value is slightly different between the two. I saw this same symptom before pulling your latest changes on the corresponding line in the old `llama-kv-cache.cpp`, so it seems that somehow the `pos` value may be getting corrupted or at least be inconsistently populated with/without `-pps`.

compilade committed:
llama-graph : fix recurrent state copy

The `state_copy` shuffle assumes everything is moved at once,
which is not true when `states_extra` is copied back to the cache
before copying the range of states between `head` and `head + n_seqs`.
This is only a problem if any of the cells in [`head`, `head + n_seqs`)
have an `src` in [`head + n_seqs`, `head + n_kv`),
which does happen when `n_ubatch > 1` in the `llama-parallel` example.

Changing the order of the operations avoids the potential overwrite
before use, although when copies are avoided (like with Mamba2),
this will require further changes.

* llama-graph : rename n_state to state_size in build_recurrent_state

This naming should reduce confusion between the state size
and the number of states.

compilade commented:
> When I start to see the with / without results diverge, the consistent thing I'm seeing is that the `pos` value is slightly different between the two. I saw this same symptom before pulling your latest changes on the corresponding line in the old `llama-kv-cache.cpp`, so it seems that somehow the `pos` value may be getting corrupted or at least be inconsistently populated with/without `-pps`.

@gabe-l-hart Interesting! The `pos` internally is there to keep track of where in the sequence the tokens come from (and how many tokens were processed), but it's not used directly in the inference (`build_inp_pos` is not used for Mamba). It's weird that this corruption somehow is not caught by the check in

<https://github.com/ggml-org/llama.cpp/blob/dd6495ddc9db4d3ba7afc0db48b581732fedb5fe/src/llama-kv-cache-recurrent.cpp#L574-L579> 

> dumping out the `cell` struct at [llama-kv-cache-recurrent.cpp:158](https://github.com/ggml-org/llama.cpp/blob/compilade/readonly-recurrent-inputs/src/llama-kv-cache-recurrent.cpp#L158).

This is in `seq_rm`, so I would assume this is happens at the end of a sequence, and if the contents differ, so might the lengths (and the `pos`).

---

Oh my, **I might have found how the problem happens**. Thank you for further narrowing what the cause could be. It really did help. The metadata *seems* to looks fine, but how the states are used is not (because it leads to different output).

I was wondering why the problem did not trigger anymore at `-ub 1`, and then I remembered [`llm_graph_context::build_recurrent_state`](https://github.com/ggml-org/llama.cpp/blob/dd6495ddc9db4d3ba7afc0db48b581732fedb5fe/src/llama-graph.cpp#L1425) now sometime makes an extra `ggml_cpy` when `n_kv > n_seqs` (usually when needing to move states around, and when more are moved than used). This does not happen at `-ub 1`, since the states never have to be moved to get 1 contiguous slot. (it would also not happen with single-sequence batches, because only one slot per sequence is needed for the recurrent cache).

More precisely, the problem only happens when a slot has `n_kv > n_seqs`, *and* the cells between `head` and `head + n` have an `src` between `head + n_seqs` and `head + n`.

<https://github.com/ggml-org/llama.cpp/blob/dd6495ddc9db4d3ba7afc0db48b581732fedb5fe/src/llama-graph.cpp#L1438-L1457>

The `state_copy` shuffles assume the shuffle happens all at once, and that is not true when `states_extra` is copied to the source states buffer. A solution is to first perform the (copying) gathering operation on the states which will be used (with `ggml_get_rows`), and *only after that* copy back the extra states to the cache.

(on `master`, the problem is caused by something else because that logic isn't there, and it seems like applying only the `swap tails` fix is not sufficient. I did not investigate further.)

With the new changes in <https://github.com/ggml-org/llama.cpp/pull/13834/commits/62a9f34baefc657212dea8f1bad14d4fb1657da8>, I'm getting deterministic multi-user output!

Mamba2 (in #9126) uses the `avoid_copies` parameter of `build_recurrent_states`, though, and the fix might need to involve a lambda function passed instead of `avoid_copies` to make sure the data is used before being potentially overwritten.

gabe-l-hart commented:
Hooray! So glad you found it. I'll test it on my side tomorrow for another set of eyes. Some day, I want to fully unpack all the moving parts so I can read this explanation fully.

ggerganov reviewed (approved):
@compilade Feel free to merge.

gabe-l-hart commented:
Confirmed on my end. With that latest fix, I see both deterministic responses when the same prompt is repeated within a parallel batch, and deterministic responses between running with and without `-pps`.
